### PR TITLE
Win ci ii

### PR DESCRIPTION
**Pull request by kwryankrattiger**
Rebased version of old Windows CI branch.

Needs #42727 
Needs #39739

### CODE REVIEW

**Comment by kwryankrattiger** [comment]
CC: @johnwparent This is more an updated reference to try to get a view on where things are at and what still need to be ported yet or re-written.

File: var/spack/repos/builtin/packages/python/package.py [hunk 1]
```diff
@@ -836,6 +836,7 @@ def command(self):
             else:
                 path = os.path.join(self.prefix, "python{0}.exe".format(ver))
             if os.path.exists(path):
+                path = f"""\"{path}\""""
```
Comment (kwryankrattiger) [inline var/spack/repos/builtin/packages/python/package.py]: Not sure if this is still needed or not.

File: share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml [hunk 2]
```diff
@@ -156,10 +156,11 @@ default:
...
       ${CI_STACK_CONFIG_SCOPES}
       audit configs
-    - spack python -c "import os,sys; print(os.path.expandvars(sys.stdin.read()))"
-      < "${SPACK_CI_CONFIG_ROOT}/${PIPELINE_MIRROR_TEMPLATE}" > "${SPACK_CI_CONFIG_ROOT}/mirrors.yaml"
+    - python ${SPACK_CI_SCRIPTS_ROOT}/common/expand_vars.py
+      "${SPACK_CI_CONFIG_ROOT}/${PIPELINE_MIRROR_TEMPLATE}"
+      "${SPACK_CI_CONFIG_ROOT}/mirrors.yaml"
     - spack config add -f "${SPACK_CI_CONFIG_ROOT}/mirrors.yaml"
-    - mkdir -p "${CI_PROJECT_DIR}/jobs_scratch_dir"
+    - mkdir "${CI_PROJECT_DIR}/jobs_scratch_dir"
```
Comment (kwryankrattiger) [inline share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml:156]: `-p` shouldn't be required, but should make sure for other platforms.

File: share/spack/gitlab/cloud_pipelines/configs/win64/ci.yaml [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+      before_script::
+      - .\bin\spack_pwsh.ps1
+      - spack compiler find
+      script::
+      - spack env activate --without-view ${SPACK_CONCRETE_ENV_DIR}
+      - spack config add "config:install_tree:projections:${SPACK_JOB_SPEC_PKG_NAME}:'morepadding/{architecture}/{compiler.name}-{compiler.version}/{name}-{version}-{hash}'"
+      - md ${SPACK_ARTIFACTS_ROOT}/user_data
+      - spack --backtrace ci rebuild > >(tee ${SPACK_ARTIFACTS_ROOT}/user_data/pipeline_out.txt) 2> >(tee ${SPACK_ARTIFACTS_ROOT}/user_data/pipeline_err.txt >&2)
+  - signing-job-remove:
+      script:: no-signing-on-windows
```
Comment (kwryankrattiger) [inline share/spack/gitlab/cloud_pipelines/configs/win64/ci.yaml]: This disables signing and it seems like so long as we don't have gpg keys available nothing should be signed.

TODO: add signing for windows binaries.

Comment (johnwparent) [reply share/spack/gitlab/cloud_pipelines/configs/win64/ci.yaml]: gpg is available on the runner image, should we setup some keys? Or do you and I have not have correct permissions on the Spack keyserver? (do we even have a keyserver?)

Comment (kwryankrattiger) [reply share/spack/gitlab/cloud_pipelines/configs/win64/ci.yaml]: If we have gpg I would be interested to see how that works. I don't know if we mount the existing intermediate key already for the protected windows runners or not, I am not sure what kind of nuance that requires.

File: share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml [hunk 4]
```diff
@@ -213,6 +220,13 @@ default:
   extends: [ ".generate-base" ]
   tags: ["spack", "public", "medium", "neoverse_v2"]
 
+.generate-win64:
+  extends: [ ".generate-common" ]
+  before_script:
+    - . .\share\spack\spack-env.ps1
```
Comment (mvandenburgh) [inline share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]:
```suggestion
    - . .\share\spack\setup-env.ps1
```

File: share/spack/gitlab/cloud_pipelines/configs/win64/ci.yaml [hunk 5]
```diff
@@ -0,0 +1,16 @@
+ci:
+  pipeline-gen:
+  - build-job:
+      after_script::
+      - Write-Output "Done"
+      before_script::
+      - . .\share\spack\spack-env.ps1
```
Comment (mvandenburgh) [inline share/spack/gitlab/cloud_pipelines/configs/win64/ci.yaml]:
```suggestion
      - . .\share\spack\setup-env.ps1
```

File: share/spack/gitlab/cloud_pipelines/scripts/common/expand_vars.py [hunk 6]
```diff
@@ -0,0 +1,10 @@
+import argparse
+import os
+
+parser = argparse.ArgumentParser()
+parser.add_argument("input", type=argparse.FileType("r"))
+parser.add_argument("out", type=argparse.FileType("w"))
+
+args = parser.parse_args()
+
+args.out.write(os.path.expandvars(args.intput.read()))
```
Comment (mvandenburgh) [inline share/spack/gitlab/cloud_pipelines/scripts/common/expand_vars.py]:
```suggestion
args.out.write(os.path.expandvars(args.input.read()))
```

File: lib/spack/spack/cmd/ci.py [hunk 7]
```diff
@@ -38,6 +38,10 @@ def deindent(desc):
     return desc.replace("    ", "")
 
 
+def encode_path(path: str) -> str:
+    return path.encode("unicode-escape".decode())
```
Comment (kwryankrattiger) [inline lib/spack/spack/cmd/ci.py]: @johnwparent Am I missing something about these lines from the cherry-pick?

```
  File "C:\builds\spack\spack\lib\spack\spack\cmd\ci.py", line 584, in ci_rebuild
    [SPACK_COMMAND, "-e", encode_path(env.path), "bootstrap", "now"],
                          ^^^^^^^^^^^^^^^^^^^^^
  File "C:\builds\spack\spack\lib\spack\spack\cmd\ci.py", line 42, in encode_path
    return path.encode("unicode-escape".decode())
                       ^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'str' object has no attribute 'decode'. Did you mean: 'encode'?
```

Comment (johnwparent) [reply lib/spack/spack/cmd/ci.py]: ~Yeah I missed the commit where I patch that typo, should be 8ff51a4fc50405be31b7b8bc6834f82a3ca16620~

You made the same change before I could finish commenting, disregard.

**Comment by kwryankrattiger** [comment]
Pulled in the latest script updates from #39739
Added escaping for `(` and `)` on windows
fixup: swap vtk for bzip2

**Comment by kwryankrattiger** [comment]
This work has been completed in another set of PRs.

**Pull request closed without merging**